### PR TITLE
Merge

### PR DESCRIPTION
**Pull request by cosa65**
# Description

# Details
#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/pipeline/pull/312
https://github.com/hms-dbmi-cellenics/worker/pull/307
https://github.com/hms-dbmi-cellenics/api/pull/442
https://github.com/hms-dbmi-cellenics/ui/pull/848

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 1]
```diff
@@ -48,23 +49,28 @@ copy_cell_sets <- function(from_experiment_id, to_experiment_id, sample_ids_map,
...
+    copy_sample_id <- sample_ids_map[[original_sample_id]]
 
-  s3_copy_by_prefix(
-    pipeline_config$source_bucket,
-    from_experiment_id,
-    pipeline_config$source_bucket,
-    to_experiment_id,
-    pipeline_config$aws_config,
-    key_transform = replace_sample_ids
-  )
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:67]: Explanation for alex: these copy changes are a fix to the existing copy which wasn't working correctly (the source and cell sets weren't being copied correctly)

**Review by alexvpickering — CHANGES_REQUESTED** [review]
see comments - mostly looks GTG

**Review by alexvpickering — APPROVED** [review]
looks good

**Review by alexvpickering — CHANGES_REQUESTED** [review]
there is an error in staging for jobs run in batch that seems to be related to metric collection for datadog (not used by HMS). processes related to datadog should not be run in these cases. Here are the error logs from batch (see bold):


> 2023-07-25T23:02:58.091Z
> 18 multi_assign(substitute(x), value, parent.frame())
> 2023-07-25T23:02:58.091Z
> 19 init-functions.R#247: run_pipeline_step(prev_out, input, pipeline_config, t
> 2023-07-25T23:02:58.091Z
> 20 init-functions.R#216: task(input, pipeline_config, prev_out)
> 2023-07-25T23:02:58.091Z
> 21 gem2s-2-load_user_files.R#37: read_fun(config, input_dir)
> 2023-07-25T23:02:58.091Z
> 22 gem2s-2-load_user_files.R#90: format_annot(annot_list)
> 2023-07-25T23:02:58.091Z
> 23 gem2s-2-load_user_files.R#327: `colnames<-`(`*tmp*`, value = c("input", "na
> 2023-07-25T23:02:58.091Z
> 24 .handleSimpleError(function (c)
> 2023-07-25T23:02:58.091Z
> {
> 2023-07-25T23:02:58.091Z
> if (inherits(c, "condition") &
**> 2023-07-25T23:02:49.211Z
> return self._submit_metrics_endpoint.call_with_http_info(**kwargs)
> 2023-07-25T23:02:49.211Z
> File "/usr/local/lib/python3.10/dist-packages/datadog_api_client/api_client.py", line 707, in call_with_http_info
> 2023-07-25T23:02:49.211Z
> self.update_params_for_auth(params["header"], params["query"])
> 2023-07-25T23:02:49.211Z
> File "/usr/local/lib/python3.10/dist-packages/datadog_api_client/api_client.py", line 742, in update_params_for_auth
> 2023-07-25T23:02:49.211Z
> raise ApiValueError("Invalid authentication token for {}".format(auth_setting["key"]))
> 2023-07-25T23:02:49.211Z
> datadog_api_client.exceptions.ApiValueError: Invalid authentication token for DD-API-KEY**
> **2023-07-25T23:02:49.210Z
> Datadog monitoring subprocess launched**
> **2023-07-25T23:02:49.210Z
> Getting instance metadata...
> 2023-07-25T23:02:49.210Z
> Collecting data every 1s, reporting to Datadog every 15s**
> **2023-07-25T23:02:49.210Z
> Traceback (most recent call last):
> 2023-07-25T23:02:49.210Z
> File "/src/pipeline-runner/./datadog-batch/monitor.py", line 88, in <module>**
> **2023-07-25T23:02:49.210Z
> response = api_instance.submit_metrics(body=payload)
> 2023-07-25T23:02:49.210Z
> File "/usr/local/lib/python3.10/dist-packages/datadog_api_client/v2/api/metrics_api.py", line 720, in submit_metrics**
> 2023-07-25T23:02:47.872Z
> Genome matrix has multiple modalities, returning a list of matrices for this genome
> 2023-07-25T23:02:46.784Z
> Genome matrix has multiple modalities, returning a list of matrices for this genome
> 2023-07-25T23:02:45.948Z
> Reading files from /input/f51f56d0-3924-4161-9e9f-3ae402f44361 --> matrix.h5.gz
> 2023-07-25T23:02:45.945Z
> Sample --> f51f56d0-3924-4161-9e9f-3ae402f44361
> 2023-07-25T23:02:45.658Z
> Genome matrix has multiple modalities, returning a list of matrices for this genome
> 2023-07-25T23:02:45.229Z
> Genome matrix has multiple modalities, returning a list of matrices for this genome
> 2023-07-25T23:02:45.017Z
> Reading files from /input/e152e344-8b56-4e5f-a35a-c63dfcf4681b --> matrix.h5.gz
> 2023-07-25T23:02:45.015Z
> Sample --> e152e344-8b56-4e5f-a35a-c63dfcf4681b

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 2]
```diff
@@ -44,6 +45,52 @@ load_user_files <- function(input, pipeline_config, prev_out, input_dir = INPUT_
...
+            " --> ",
+            paste(sample_fpaths, collapse = " - "))
+
+    if (length(sample_fpaths) > 1)
+      stop("Only one h5 expected. More files detected.")
+
+    ungzipped_counts_path <- R.utils::gunzip(sample_counts_path)
+
+    gene_names <- rownames(Seurat::Read10X_h5(ungzipped_counts_path))
+    counts <- Seurat::Read10X_h5(ungzipped_counts_path, use.names = F)
```
Comment (alexvpickering) [inline pipeline-runner/R/gem2s-2-load_user_files.R:82]: This fixes a bug where an H5 pipeline will fail if there are multiple modalities. To test, you can use [these H5 files](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE183716) which are from the workshop dataset

```suggestion
    counts_names <- Seurat::Read10X_h5(ungzipped_counts_path)
    counts <- Seurat::Read10X_h5(ungzipped_counts_path, use.names = FALSE)

    # use Gene Expression modality if multiple
    if (methods::is(counts, 'list')) {
      counts_names <- counts_names$`Gene Expression`
      counts <- counts$`Gene Expression`
    }

    gene_names <- row.names(counts_names)
```

**Review by alexvpickering — APPROVED** [review]

**Pull request merged**